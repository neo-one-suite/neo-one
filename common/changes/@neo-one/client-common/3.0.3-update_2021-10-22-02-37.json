{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Update nef model and types",
      "type": "minor"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "29364457+spencercorwin@users.noreply.github.com"
}
