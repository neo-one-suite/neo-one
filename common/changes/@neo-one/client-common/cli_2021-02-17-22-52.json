{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Fix minor bugs.",
      "type": "patch"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "spencercorwin@icloud.com"
}
