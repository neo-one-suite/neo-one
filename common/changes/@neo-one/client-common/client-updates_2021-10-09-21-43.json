{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Add serializeJSON to attributes",
      "type": "patch"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "29364457+spencercorwin@users.noreply.github.com"
}
