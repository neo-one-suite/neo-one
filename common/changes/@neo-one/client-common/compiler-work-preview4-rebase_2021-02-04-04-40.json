{
  "changes": [
    {
      "packageName": "@neo-one/client-common",
      "comment": "Update compiler, client, and CLI for Preview4",
      "type": "patch"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "spencercorwin@icloud.com"
}
