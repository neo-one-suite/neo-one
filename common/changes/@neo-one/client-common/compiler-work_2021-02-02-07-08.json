{
  "changes": [
    {
      "comment": "Update for Preview3, primarily in compiler, client, and CLI.",
      "type": "patch",
      "packageName": "@neo-one/client-common"
    }
  ],
  "packageName": "@neo-one/client-common",
  "email": "spencercorwin@icloud.com"
}
