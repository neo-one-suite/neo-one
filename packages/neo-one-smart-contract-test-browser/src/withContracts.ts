import { NEOONEDataProvider } from '@neo-one/client-core';
import { createCompilerHost, FileSystem } from '@neo-one/local-browser';
import { createJSONRPCLocalProviderManager, getFileSystem } from '@neo-one/local-singleton';
import { JSONRPCLocalProvider } from '@neo-one/node-browser';
import {
  Contract,
  TestOptions,
  withContracts as withContractsBase,
  WithContractsOptions,
} from '@neo-one/smart-contract-test-common';
import { constants } from '@neo-one/utils';
import { WorkerManager } from '@neo-one/worker';

export { TestOptions, WithContractsOptions, Contract };

export const createWithContractsBase =
  (getFS: () => FileSystem, createManager: () => Promise<WorkerManager<typeof JSONRPCLocalProvider>>) =>
  async <T>(
    contracts: ReadonlyArray<Contract>,
    test: (contracts: T & TestOptions) => Promise<void>,
    options?: WithContractsOptions,
  ): Promise<void> =>
    withContractsBase(
      contracts,
      test,
      () => createCompilerHost({ fs: getFS() }),
      async () => {
        const manager = await createManager();
        const dataProvider = new NEOONEDataProvider({ network: 'priv', rpcURL: manager });

        return {
          dataProvider,
          privateKey: constants.PRIVATE_NET_PRIVATE_KEY,
          cleanup: async () => {
            manager.dispose();
          },
        };
      },
      options,
    );

export const withContracts = createWithContractsBase(getFileSystem, createJSONRPCLocalProviderManager);
