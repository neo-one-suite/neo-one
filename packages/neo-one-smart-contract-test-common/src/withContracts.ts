import { setupWallets } from '@neo-one/cli-common';
import {
  assertCallFlags,
  common,
  crypto,
  MethodTokenModel,
  NefFileModel,
  scriptHashToAddress,
  SignerModel,
  SourceMaps,
  TransactionResult,
  UserAccountID,
  WitnessScopeModel,
} from '@neo-one/client-common';
import { DeveloperClient, LocalKeyStore, NEOONEDataProvider, NEOONEProvider } from '@neo-one/client-core';
import { Client, LocalUserAccountProvider, PublishReceipt } from '@neo-one/client-full-core';
import { setGlobalLogLevel } from '@neo-one/logger';
import { compileContract, CompilerHost } from '@neo-one/smart-contract-compiler';
import { camel, Modifiable } from '@neo-one/utils';
import BigNumber from 'bignumber.js';

export interface Contract {
  readonly filePath: string;
  readonly name: string;
}
export interface WithContractsOptions {
  /**
   * Ignore compiler warnings. Useful during smart contract development.
   *
   * Defaults to `false`.
   */
  readonly ignoreWarnings?: boolean;
  /**
   * Automatically deploy smart contracts using the defaults specified in the constructor arguments.
   *
   * Defaults to `true`.
   */
  readonly deploy?: boolean;
  /**
   * Automatically run consensus whenever a transaction is relayed.
   *
   * Defaults to `true`.
   */
  readonly autoConsensus?: boolean;
  /**
   * Automatically provide the necessary system fee for every transaction to execute.
   *
   * Defaults to `true`.
   */
  readonly autoSystemFee?: boolean;
  /**
   * Enable logs from various systems during testing.
   *
   * Defaults to `false`.
   */
  readonly logging?: boolean;
}

/**
 * `TestOptions` holds all of the properties that are pre-configured and available for use in tests.
 *
 * In addition to the properties listed below, the object will contain a smart contract API object property for each smart contract in your project.
 */
export interface TestOptions {
  /**
   * The local network name that the smart contracts have been deployed to and the `client` has been configured with.
   */
  readonly networkName: string;
  /**
   * `Client` that has been pre-configured with the master account for the local network as well as each of the accounts in `accountIDs`.
   *
   * This user account is also the currently selected user account in the `Client` and the one that deployed the contracts.
   */
  readonly client: Client<{
    readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>;
  }>;
  /**
   * `DeveloperClient` that's been configured to point at the local testing network.
   */
  readonly developerClient: DeveloperClient;
  /**
   * `UserAccountID` of the "master" account - the account that contains ~100 million NEO and ~58 million GAS.
   *
   * This user account is also the currently selected user account in the `Client`.
   */
  readonly masterAccountID: UserAccountID;
  /**
   * Private key for the `masterAccountID`.
   */
  readonly masterPrivateKey: string;
  /**
   * 10 additional user accounts that have been configured in the client with varying amounts of NEO and GAS:
   *
   * At index:
   *  0. 0 NEO and GAS
   *  1. 1 NEO and GAS
   *  2. 10 NEO and GAS
   *  3. 100 NEO and GAS
   *  4. 1000 NEO and GAS
   *  5. 10000 NEO and GAS
   *  6. 100000 NEO and GAS
   *  7. 1000000 NEO and GAS
   *  8. 5 NEO and GAS
   *  9. 20 NEO and GAS
   */
  readonly accountIDs: ReadonlyArray<UserAccountID>;
}

export interface DataProviderOptions {
  readonly dataProvider: NEOONEDataProvider;
  readonly cleanup: () => Promise<void>;
  readonly privateKey: string;
}

export const withContracts = async <T>(
  contracts: ReadonlyArray<Contract>,
  test: (contracts: T & TestOptions) => Promise<void>,
  createCompilerHost: () => CompilerHost,
  getDataProvider: () => Promise<DataProviderOptions>,
  {
    ignoreWarnings = false,
    deploy = true,
    autoConsensus = true,
    autoSystemFee = true,
    logging = false,
  }: WithContractsOptions = {},
): Promise<void> => {
  setGlobalLogLevel(logging ? 'info' : 'silent');
  const { dataProvider, cleanup, privateKey } = await getDataProvider();
  const { client, developerClient, masterWallet, accountIDs } = await setupWallets(dataProvider, privateKey);
  const networkName = dataProvider.network;
  try {
    if (autoSystemFee) {
      client.hooks.beforeRelay.tapPromise('AutoSystemFee', async (options) => {
        // tslint:disable-next-line no-object-mutation
        options.maxSystemFee = new BigNumber(-1);
        // tslint:disable-next-line no-object-mutation
        options.maxNetworkFee = new BigNumber(-1);
      });
    }
    if (autoConsensus) {
      client.hooks.beforeConfirmed.tapPromise('DeveloperClient', async () => {
        await developerClient.runConsensusNow();
      });
    }
    const mutableLinked: { [filePath: string]: { [contractName: string]: string } } = {};
    const mutableSourceMaps: Modifiable<SourceMaps> = {};

    const deployedContracts = await contracts.reduce<Promise<T>>(async (accIn, { filePath, name }) => {
      const acc = await accIn;
      const { contract, sourceMap } = await compileContract(
        filePath,
        name,
        createCompilerHost(),
        mutableLinked,
        ignoreWarnings,
      );

      const signerAddress = masterWallet.userAccount.id.address;
      const signer = new SignerModel({
        account: crypto.addressToScriptHash({
          address: signerAddress,
          addressVersion: common.NEO_ADDRESS_VERSION,
        }),
        scopes: WitnessScopeModel.Global,
      });

      const nefFile = new NefFileModel({
        compiler: contract.nefFile.compiler,
        script: Buffer.from(contract.nefFile.script, 'hex'),
        tokens: contract.nefFile.tokens.map(
          (token) =>
            new MethodTokenModel({
              hash: common.stringToUInt160(token.hash),
              method: token.method,
              paramCount: token.paramCount,
              hasReturnValue: token.hasReturnValue,
              callFlags: assertCallFlags(token.callFlags),
            }),
        ),
      });

      // TODO: create function for getting contract address from register/signer?
      const contractHash = crypto.getContractHash(signer.account, nefFile.checkSum, contract.manifest.name);
      const address = scriptHashToAddress(common.uInt160ToString(contractHash));

      mutableSourceMaps[address] = await sourceMap;
      let result: TransactionResult<PublishReceipt>;
      // tslint:disable-next-line prefer-conditional-expression
      if (deploy) {
        result = await client.publishAndDeploy(
          contract,
          contract.manifest,
          [signerAddress], // TODO: fix contract calling here if necessary. Probably just stays empty array here
          { maxSystemFee: new BigNumber(-1), maxNetworkFee: new BigNumber(-1) },
          mutableSourceMaps,
        );
      } else {
        result = await client.publish(contract, { maxSystemFee: new BigNumber(-1), maxNetworkFee: new BigNumber(-1) });
      }

      const [receipt] = await Promise.all([result.confirmed({ timeoutMS: 2500 }), developerClient.runConsensusNow()]);
      if (receipt.result.state === 'FAULT') {
        throw new Error(receipt.result.message);
      }

      const smartContract = client.smartContract({
        networks: { [networkName]: { address } },
        manifest: contract.manifest,
        sourceMaps: mutableSourceMaps,
      });
      mutableLinked[filePath] = { [name]: address };

      // tslint:disable-next-line prefer-object-spread
      return Object.assign({}, acc, { [camel(name)]: smartContract }) as T;
      // tslint:disable-next-line no-any
    }, Promise.resolve({} as any));

    // tslint:disable-next-line:prefer-object-spread
    const contractOptions = Object.assign({}, deployedContracts, {
      client,
      developerClient,
      masterAccountID: masterWallet.userAccount.id,
      masterPrivateKey: privateKey,
      networkName,
      accountIDs,
    });

    await test(contractOptions);
  } finally {
    await cleanup();
  }
};
