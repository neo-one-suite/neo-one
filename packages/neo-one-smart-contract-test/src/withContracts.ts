import {
  Client,
  DeveloperClient,
  LocalKeyStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  PublishReceipt,
  scriptHashToAddress,
  SourceMaps,
  TransactionResult,
  UserAccountID,
} from '@neo-one/client';
import { common, crypto } from '@neo-one/client-core';
import { compileContract } from '@neo-one/smart-contract-compiler';
import { camel } from 'change-case';
import { setupTestNode } from './setupTestNode';

export interface Contract {
  readonly filePath: string;
  readonly name: string;
}
export interface WithContractsOptions {
  readonly ignoreWarnings?: boolean;
  readonly deploy?: boolean;
}

export interface TestOptions {
  readonly networkName: string;
  readonly client: Client<{
    readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>;
  }>;
  readonly developerClient: DeveloperClient;
  readonly masterAccountID: UserAccountID;
  readonly masterPrivateKey: string;
}

export const withContracts = async <T>(
  contracts: ReadonlyArray<Contract>,
  test: (contracts: T & TestOptions) => Promise<void>,
  { ignoreWarnings = false, deploy = true }: WithContractsOptions = { ignoreWarnings: false, deploy: true },
): Promise<void> => {
  const { client, masterWallet, provider, networkName, privateKey, node } = await setupTestNode();
  try {
    const developerClient = new DeveloperClient(provider.read(networkName));
    const mutableLinked: { [filePath: string]: { [contractName: string]: string } } = {};
    const mutableSourceMaps: Modifiable<SourceMaps> = {};

    const deployedContracts = await contracts.reduce<Promise<T>>(async (accIn, { filePath, name }) => {
      const acc = await accIn;
      const { contract, sourceMap, abi } = compileContract(filePath, name, mutableLinked, ignoreWarnings);
      const address = scriptHashToAddress(
        common.uInt160ToString(crypto.toScriptHash(Buffer.from(contract.script, 'hex'))),
      );
      mutableSourceMaps[address] = await sourceMap;
      let result: TransactionResult<PublishReceipt>;
      // tslint:disable-next-line prefer-conditional-expression
      if (deploy) {
        result = await client.publishAndDeploy(contract, abi, [], undefined, mutableSourceMaps);
      } else {
        result = await client.publish(contract);
      }

      const [receipt] = await Promise.all([result.confirmed({ timeoutMS: 2500 }), developerClient.runConsensusNow()]);
      if (receipt.result.state === 'FAULT') {
        throw new Error(receipt.result.message);
      }

      const smartContract = client.smartContract({
        networks: { [networkName]: { address } },
        abi,
        sourceMaps: mutableSourceMaps,
      });
      mutableLinked[filePath] = { [name]: address };

      // tslint:disable-next-line prefer-object-spread
      return Object.assign({}, acc, { [camel(name)]: smartContract }) as T;
      // tslint:disable-next-line no-any
    }, Promise.resolve({} as any));

    // tslint:disable-next-line:prefer-object-spread
    const contractOptions = Object.assign({}, deployedContracts, {
      client,
      developerClient,
      masterAccountID: masterWallet.account.id,
      masterPrivateKey: privateKey,
      provider,
      networkName,
    });

    await test(contractOptions);
  } finally {
    await node.stop();
  }
};
