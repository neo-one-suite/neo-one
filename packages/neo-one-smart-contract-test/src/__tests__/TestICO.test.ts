import { common, crypto, privateKeyToAddress } from '@neo-one/client-common';
import { Hash256, SmartContractAny } from '@neo-one/client-core';
import BigNumber from 'bignumber.js';
import * as path from 'path';
import { withContracts } from '../withContracts';

const MINTER = {
  PRIVATE_KEY: '536f1e9f0466f6cd5b2ea5374d00f038786daa0f0e892161d6b0cb4d6b154740',
  PUBLIC_KEY: '03463b7a0afc41ff1f6a386190f99bafd1deca48f4026aeac95435731af278cb7d',
};

describe('TestICO', () => {
  test('smart contract', async () => {
    await withContracts<{ testIco: SmartContractAny }>(
      [
        {
          filePath: path.resolve(__dirname, '..', '__data__', 'contracts', 'TestICO.ts'),
          name: 'TestICO',
        },
      ],
      async ({ client, networkName, testIco: smartContract, masterAccountID }) => {
        crypto.addPublicKey(common.stringToPrivateKey(MINTER.PRIVATE_KEY), common.stringToECPoint(MINTER.PUBLIC_KEY));
        const deployResult = await smartContract.deploy(
          masterAccountID.address,
          new BigNumber(Math.round(Date.now() / 1000)),
        );
        const deployReceipt = await deployResult.confirmed({ timeoutMS: 2500 });
        if (deployReceipt.result.state !== 'HALT') {
          throw new Error(deployReceipt.result.message);
        }

        expect(deployReceipt.result.gasConsumed.toString()).toMatchSnapshot('deploy consumed');
        expect(deployReceipt.result.gasCost.toString()).toMatchSnapshot('deploy cost');
        expect(deployReceipt.result.value).toBeTruthy();

        const [nameResult, decimalsResult, symbolResult, minter] = await Promise.all([
          smartContract.name(),
          smartContract.decimals(),
          smartContract.symbol(),
          client.providers.memory.keystore.addUserAccount({
            network: networkName,
            name: 'minter',
            privateKey: MINTER.PRIVATE_KEY,
          }),
        ]);
        expect(nameResult).toEqual('TestToken');
        expect(decimalsResult.toString()).toEqual('8');
        expect(symbolResult).toEqual('TT');

        const [initialTotalSupply, transferResult] = await Promise.all([
          smartContract.totalSupply(),
          client.transfer(
            [
              {
                amount: new BigNumber(10000),
                asset: Hash256.NEO,
                to: privateKeyToAddress(MINTER.PRIVATE_KEY),
              },
              {
                amount: new BigNumber(10000),
                asset: Hash256.GAS,
                to: privateKeyToAddress(MINTER.PRIVATE_KEY),
              },
            ],
            {
              from: masterAccountID,
            },
          ),
        ]);
        expect(initialTotalSupply.toString()).toEqual('0');

        await transferResult.confirmed({ timeoutMS: 2500 });

        const firstMint = new BigNumber('10');
        const mintResult = await smartContract.mintTokens({
          from: minter.userAccount.id,
          sendTo: [
            {
              amount: firstMint,
              asset: Hash256.NEO,
            },
          ],
        });

        const mintReceipt = await mintResult.confirmed({ timeoutMS: 2500 });
        expect(mintReceipt.result.gasConsumed.toString()).toMatchSnapshot('mint consumed');
        expect(mintReceipt.result.gasCost.toString()).toMatchSnapshot('mint cost');
        expect(mintReceipt.result.value).toBeUndefined();
        expect(mintReceipt.events).toHaveLength(1);
        const event = mintReceipt.events[0];
        expect(event.name).toEqual('transfer');
        expect(event.parameters.from).toBeUndefined();
        expect(event.parameters.to).toEqual(minter.userAccount.id.address);
        if (event.parameters.amount === undefined) {
          expect(event.parameters.amount).toBeTruthy();
          throw new Error('For TS');
        }
        const firstBalance = firstMint.times(10).toString();
        expect(event.parameters.amount.toString()).toEqual(firstBalance);

        const [minterBalance, mintTotalSupply] = await Promise.all([
          smartContract.balanceOf(minter.userAccount.id.address),
          smartContract.totalSupply(),
        ]);

        expect(minterBalance.toString(10)).toEqual(firstBalance);
        expect(mintTotalSupply.toString(10)).toEqual(firstBalance);
      },
      { deploy: false },
    );
  });
});
