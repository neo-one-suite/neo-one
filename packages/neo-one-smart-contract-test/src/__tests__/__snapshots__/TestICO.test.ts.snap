// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestICO smart contract: deploy consumed 1`] = `"0"`;

exports[`TestICO smart contract: deploy cost 1`] = `"1.494"`;

exports[`TestICO smart contract: mint consumed 1`] = `"0"`;

exports[`TestICO smart contract: mint cost 1`] = `"4.809"`;
